bundles/org.openhab.binding.enigma2/src/main/java/org/openhab/binding/enigma2/handler/Enigma2Handler.java

```diff
@@ -0,0 +1,305 @@
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.openhab.binding.enigma2.actions.Enigma2Actions;
+import org.openhab.binding.enigma2.internal.Enigma2Client;
+import org.openhab.binding.enigma2.internal.Enigma2Configuration;
+import org.openhab.binding.enigma2.internal.Enigma2RemoteKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The {@link Enigma2Handler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Guido Dolfen - Initial contribution
+ */
+@NonNullByDefault
+public class Enigma2Handler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(Enigma2Handler.class);
+    private Enigma2Configuration config;
+    private Enigma2Client enigma2Client;
+    @Nullable
+    private ScheduledFuture<?> refreshJob;
+
+    public Enigma2Handler(Thing thing) {
+        super(thing);
+        config = getConfigAs(Enigma2Configuration.class);
+        enigma2Client = new Enigma2Client(config.host, config.user, config.password, config.timeout);
+    }
+
+    @Override
+    public void initialize() {
+        refreshJob = scheduler.scheduleWithFixedDelay(this::refresh, 2, config.refreshInterval, TimeUnit.SECONDS);
+    }
+
+    private void refresh() {
+        try {
+            boolean online = getEnigma2Client().refresh();
+            if (online) {
+                updateStatus(ThingStatus.ONLINE);
+                updateState(CHANNEL_POWER, getEnigma2Client().isPower() ? OnOffType.ON : OnOffType.OFF);
+                updateState(CHANNEL_MUTE, getEnigma2Client().isMute() ? OnOffType.ON : OnOffType.OFF);
+                updateState(CHANNEL_VOLUME, new PercentType(getEnigma2Client().getVolume()));
+                updateState(CHANNEL_CHANNEL, new StringType(getEnigma2Client().getChannel()));
+                updateState(CHANNEL_TITLE, new StringType(getEnigma2Client().getTitle()));
+                updateState(CHANNEL_DESCRIPTION, new StringType(getEnigma2Client().getDescription()));
+            } else {
+                updateStatus(ThingStatus.OFFLINE);
+            }
+        } catch (Exception e) {
+            updateStatus(ThingStatus.UNINITIALIZED);
+            getLogger().warn("Error on refreshing {}", thing.getUID(), e);
+        }
+    }
+
+    @Override
+    @SuppressWarnings("null")
+    public void handleRemoval() {
+        refreshJob.cancel(true);
```

Comment:
Please move the job canceling into the `dispose` method, that is the better place according the the ThingHandler lifecycle. If you have moved it to "dispose", please also do

```
if(refreshJob != null) {
   if(!refreshJob.isCanceled()) {
        refreshJob.cancel(true);
   }
   refreshJob = null;
}
```

Also: This might be a possible null pointer exception, please check `refreshJob != null` before accessing it.

Reply:
Fixed